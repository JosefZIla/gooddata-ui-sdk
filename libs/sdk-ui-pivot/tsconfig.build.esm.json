{
    "extends": "./tsconfig.build.json",
    "include": ["src/index.ts", "src/typings.d.ts"],
    "compilerOptions": {
        "module": "ESNext",
        "outDir": "esm",
        "incremental": true
    }
}
