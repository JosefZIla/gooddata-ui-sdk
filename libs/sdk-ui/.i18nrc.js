module.exports = {
    paths: ["./src/base/localization/bundles"],
    structure: true,
    intl: true,
    html: true,
    insightToReport: true,
    usage: false,
    debug: false,
    source: "src/**/*.{ts,js,tsx,jsx}",
    rules: [
        {
            pattern: [/.+/],
        },
    ],
};
