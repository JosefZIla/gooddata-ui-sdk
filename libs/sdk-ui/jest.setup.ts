// (C) 2019 GoodData Corporation
/* eslint-disable import/no-unassigned-import */
import * as raf from "raf";
import "@testing-library/jest-dom/extend-expect";

raf.polyfill();
