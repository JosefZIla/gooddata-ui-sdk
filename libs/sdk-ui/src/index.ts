// (C) 2007-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
import "./polyfills";
// new exports

export * from "./base";
export * from "./execution";
export * from "./kpi";

/**
 * Common interface uses to specify number separators for the different SDK components.
 * @public
 */
export { ISeparators } from "@gooddata/sdk-backend-spi";
