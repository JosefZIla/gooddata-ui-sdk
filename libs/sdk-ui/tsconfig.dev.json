{
    "extends": "./tsconfig.build.esm.json",
    "compilerOptions": {
        "noImplicitAny": false,
        "noUnusedLocals": false,
        "noUnusedParameters": false,
        "pretty": true
    }
}
