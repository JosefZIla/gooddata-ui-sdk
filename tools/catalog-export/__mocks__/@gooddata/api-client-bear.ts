// (C) 2007-2018 GoodData Corporation
import catalogItemsResponse from "./loadItems.json";
import getObjectsResponse from "./getObjects.json";
import getObjectsByQueryResponse from "./getObjectsByQuery.json";
import getVisualizationsResponse from "./getVisualizations.json";
import getAnalyticalDashboards from "./getAnalyticalDashboards.json";

export default {
    catalogue: {
        loadItems() {
            return Promise.resolve(catalogItemsResponse);
        },
    },
    md: {
        getObjectsByQuery() {
            return Promise.resolve(getObjectsByQueryResponse);
        },
        getObjects() {
            return Promise.resolve(getObjectsResponse);
        },
        getVisualizations() {
            return Promise.resolve(getVisualizationsResponse);
        },
        getAnalyticalDashboards() {
            return Promise.resolve(getAnalyticalDashboards);
        },
    },
};
