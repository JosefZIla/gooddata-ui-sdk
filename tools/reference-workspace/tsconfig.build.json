{
    "extends": "./tsconfig.json",
    "include": ["src/index.ts", "src/typings.d.ts"],
    "compilerOptions": {
        "declaration": true,
        "declarationMap": true
    }
}
