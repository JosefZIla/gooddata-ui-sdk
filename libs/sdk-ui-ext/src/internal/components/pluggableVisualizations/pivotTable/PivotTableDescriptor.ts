// (C) 2021-2022 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import isNil from "lodash/isNil";
import {
    bucketAttributes,
    bucketMeasures,
    IInsight,
    IInsightDefinition,
    insightFilters,
    insightProperties,
    insightSanitize,
    insightSorts,
    insightTotals,
} from "@gooddata/sdk-model";
import {
    IAttributeColumnWidthItem,
    IColumnSizing,
    IPivotTableConfig,
    IPivotTableProps,
    isAttributeColumnWidthItem,
    pivotTableMenuForCapabilities,
} from "@gooddata/sdk-ui-pivot";
import { BucketNames } from "@gooddata/sdk-ui";
import { ISettings } from "@gooddata/sdk-backend-spi";

import {
    IEmbeddingCodeContext,
    IVisualizationDescriptor,
    IVisualizationSizeInfo,
    PluggableVisualizationFactory,
} from "../../../interfaces/VisualizationDescriptor";
import { IFluidLayoutDescriptor } from "../../../interfaces/LayoutDescriptor";
import { PluggablePivotTable } from "./PluggablePivotTable";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
import { IDrillDownContext } from "../../../interfaces/Visualization";
import {
    addIntersectionFiltersToInsight,
    modifyBucketsAttributesForDrillDown,
    sanitizeTableProperties,
} from "../drillDownUtil";
import { getReactEmbeddingCodeGenerator } from "../../../utils/embeddingCodeGenerator";
import {
    bucketConversion,
    getInsightToPropsConverter,
    insightConversion,
} from "../../../utils/embeddingCodeGenerator/insightToPropsConverter";

export class PivotTableDescriptor extends BaseChartDescriptor implements IVisualizationDescriptor {
    public getFactory(): PluggableVisualizationFactory {
        return (params) => new PluggablePivotTable(params);
    }

    public getSizeInfo(
        _insight: IInsightDefinition,
        layoutDescriptor: IFluidLayoutDescriptor,
        settings: ISettings,
    ): IVisualizationSizeInfo {
        return {
            width: {
                default: layoutDescriptor.gridColumnsCount,
                min: 3,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getMinHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    }

    public applyDrillDown(insight: IInsight, drillDownContext: IDrillDownContext): IInsight {
        const drillDownInsight = modifyBucketsAttributesForDrillDown(
            insight,
            drillDownContext.drillDefinition,
        );
        const drillDownInsightWithFilters = addIntersectionFiltersToInsight(
            drillDownInsight,
            drillDownContext.event.drillContext.intersection,
        );
        return sanitizeTableProperties(insightSanitize(drillDownInsightWithFilters));
    }

    public getEmbeddingCode = getReactEmbeddingCodeGenerator({
        component: {
            importType: "named",
            name: "PivotTable",
            package: "@gooddata/sdk-ui-pivot",
        },
        insightToProps: getInsightToPropsConverter<IPivotTableProps>({
            measures: bucketConversion("measures", BucketNames.MEASURES, bucketMeasures),
            rows: bucketConversion("rows", BucketNames.ATTRIBUTE, bucketAttributes),
            columns: bucketConversion("columns", BucketNames.COLUMNS, bucketAttributes),
            filters: insightConversion("filters", insightFilters),
            sortBy: insightConversion("sortBy", insightSorts),
            totals: insightConversion("totals", insightTotals),
            config: insightConversion("config", getConfigFromInsight),
        }),
        additionalFactories: [
            {
                importInfo: {
                    name: "newWidthForAttributeColumn",
                    package: "@gooddata/sdk-ui-pivot",
                    importType: "named",
                },
                transformation: (obj) => {
                    return isAttributeColumnWidthItem(obj)
                        ? factoryNotationForAttributeColumnWidthItem(obj)
                        : undefined;
                },
            },
        ],
    });
}

function getColumnSizingFromControls(
    controls: Record<string, any>,
    ctx: IEmbeddingCodeContext | undefined,
): IColumnSizing | undefined {
    const { columnWidths } = controls;
    const columnWidthsProp = !isEmpty(columnWidths) ? { columnWidths } : {};

    const growToFitConfig = ctx?.settings?.enableTableColumnsGrowToFit;
    const growToFitProp = !isNil(growToFitConfig) ? { growToFit: growToFitConfig } : {};

    return {
        ...columnWidthsProp,
        ...growToFitProp,
        // the user can fill the rest on their own later
    };
}

function getConfigFromInsight(
    insight: IInsightDefinition,
    ctx: IEmbeddingCodeContext | undefined,
): IPivotTableConfig {
    const properties = insightProperties(insight);
    const controls = properties?.controls;
    const columnSizing = controls && getColumnSizingFromControls(controls, ctx);

    const menuConfig = ctx?.backend && pivotTableMenuForCapabilities(ctx.backend.capabilities);
    const menuProp = !isEmpty(menuConfig) ? { menu: menuConfig } : {};

    const separatorsConfig = ctx?.settings?.separators;
    const separatorsProp = !isEmpty(separatorsConfig) ? { separators: separatorsConfig } : {};

    return {
        columnSizing,
        ...menuProp,
        ...separatorsProp,
        // the user can fill the rest on their own later
    };
}

function factoryNotationForAttributeColumnWidthItem(obj: IAttributeColumnWidthItem): string {
    const { attributeIdentifier, width } = obj.attributeColumnWidthItem;
    const { value: widthValue, allowGrowToFit } = width;
    return allowGrowToFit
        ? `newWidthForAttributeColumn(${attributeIdentifier}, ${widthValue}, true)`
        : `newWidthForAttributeColumn(${attributeIdentifier}, ${widthValue})`;
}
