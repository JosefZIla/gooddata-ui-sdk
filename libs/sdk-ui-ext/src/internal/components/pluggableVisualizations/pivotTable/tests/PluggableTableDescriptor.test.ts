// (C) 2022 GoodData Corporation
import { ReferenceMd } from "@gooddata/reference-workspace";
import {
    modifyAttribute,
    modifyMeasure,
    newAttributeSort,
    newBucket,
    newInsightDefinition,
    newTotal,
} from "@gooddata/sdk-model";
import { BucketNames } from "@gooddata/sdk-ui";
import { newWidthForAttributeColumn } from "@gooddata/sdk-ui-pivot";
import { PivotTableDescriptor } from "../PivotTableDescriptor";

describe("PivotTableDescriptor", () => {
    it("should generate embedding code without context", () => {
        // TODO replace by some recording?
        const insight = newInsightDefinition("local:table", (i) =>
            i
                .buckets([
                    newBucket(
                        BucketNames.MEASURES,
                        modifyMeasure(ReferenceMd.Amount, (m) =>
                            m.alias("Custom Amount alias").localId("amount-custom-local-id"),
                        ),
                    ),
                    newBucket(
                        BucketNames.ATTRIBUTE,
                        modifyAttribute(ReferenceMd.Region, (a) =>
                            a.alias("Custom Region alias").localId("region-custom-local-id"),
                        ),
                        modifyAttribute(ReferenceMd.Department, (a) =>
                            a.alias("Custom Department alias").localId("department-custom-local-id"),
                        ),
                        newTotal("sum", "amount-custom-local-id", "region-custom-local-id"),
                        newTotal("sum", "amount-custom-local-id", "department-custom-local-id"),
                    ),
                    newBucket(
                        BucketNames.COLUMNS,
                        modifyAttribute(ReferenceMd.Activity.Subject, (a) =>
                            a.alias("Custom Subject alias").localId("subject-custom-local-id"),
                        ),
                    ),
                ])
                .sorts([newAttributeSort("region-custom-local-id", "asc")])
                .properties({
                    controls: {
                        columnWidths: [
                            newWidthForAttributeColumn("region-custom-local-id", 256),
                            newWidthForAttributeColumn("department-custom-local-id", 180),
                        ],
                    },
                }),
        );
        const descriptor = new PivotTableDescriptor();

        expect(descriptor.getEmbeddingCode(insight)).toMatchSnapshot();
    });
});
