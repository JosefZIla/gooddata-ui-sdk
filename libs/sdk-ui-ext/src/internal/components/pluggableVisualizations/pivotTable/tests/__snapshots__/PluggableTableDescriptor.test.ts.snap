// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PivotTableDescriptor should generate embedding code without context 1`] = `
"import React from \\"react\\";
import { idRef, newAttribute, newAttributeSort, newMeasure, newTotal } from \\"@gooddata/sdk-model\\";
import { newWidthForAttributeColumn, PivotTable } from \\"@gooddata/sdk-ui-pivot\\";

const measures = [
    newMeasure(idRef(\\"aangOxLSeztu\\", \\"measure\\"), m => m.alias(\\"Custom Amount alias\\").localId(\\"m_aangOxLSeztu\\"))
];
const rows = [
    newAttribute(idRef(\\"label.owner.region\\", \\"displayForm\\"), a => a.alias(\\"Custom Region alias\\").localId(\\"a_label.owner.region\\")),
    newAttribute(idRef(\\"label.owner.department\\", \\"displayForm\\"), a => a.alias(\\"Custom Department alias\\").localId(\\"a_label.owner.department\\"))
];
const columns = [
    newAttribute(idRef(\\"label.activity.id.subject\\", \\"displayForm\\"), a => a.alias(\\"Custom Subject alias\\").localId(\\"a_label.activity.id.subject\\"))
];
const sortBy = [newAttributeSort(\\"a_label.owner.region\\", \\"asc\\")];
const totals = [
    newTotal(\\"sum\\", \\"m_aangOxLSeztu\\", \\"a_label.owner.region\\"),
    newTotal(\\"sum\\", \\"m_aangOxLSeztu\\", \\"a_label.owner.department\\")
];
const config = {
    columnSizing: {
        columnWidths: [
            newWidthForAttributeColumn(a_label.owner.region, 256),
            newWidthForAttributeColumn(a_label.owner.department, 180)
        ]
    }
};
const style = { height: 400 };

function MyComponent() {
    return (
        <div style={style}>
            <PivotTable
                measures={measures}
                rows={rows}
                columns={columns}
                sortBy={sortBy}
                totals={totals}
                config={config}
            />
        </div>
    );
}
"
`;
